cmd/ipam-node/main.go

```diff
@@ -13,7 +13,312 @@
+	hash := sha256.New()
+	writer := io.MultiWriter(hash, fp)
+
+	// genearate kubeconfig from template
+	if err = templateKubeconfig.Execute(writer, templateData); err != nil {
+		return nil, fmt.Errorf("cannot create kubeconfig: %v", err)
+	}
+
+	if err := fp.Sync(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot flush kubeconfig temp file: %v", err)
+	}
+	if err := fp.Close(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot close kubeconfig temp file: %v", err)
+	}
+
+	newFileHash := hash.Sum(nil)
+	if currentFileHash != nil && bytes.Equal(newFileHash, currentFileHash) {
+		fmt.Printf("kubeconfig is same, not copy\n")
+		os.Remove(fp.Name())
+		return currentFileHash, nil
+	}
+
+	// replace file with tempfile
+	if err := os.Rename(tempKubeConfigFile, nvIPAMKubeConfig); err != nil {
+		return nil, fmt.Errorf("cannot replace %q with temp file %q: %v", nvIPAMKubeConfig, tempKubeConfigFile, err)
+	}
+
+	fmt.Printf("kubeconfig is created in %s\n", nvIPAMKubeConfig)
+	return newFileHash, nil
+}
+func (o *Options) createNvIpamConfig(currentFileHash []byte) ([]byte, error) {
+	// create kubeconfig by template and replace it by atomic
+	tempNvIpamConfigFile := fmt.Sprintf("%s/nv-ipam.d/nv-ipam.conf.new", o.CNIConfDir)
+	nvIpamConfigFile := fmt.Sprintf("%s/nv-ipam.d/nv-ipam.conf", o.CNIConfDir)
+	fp, err := os.OpenFile(tempNvIpamConfigFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create nv-ipam.conf temp file: %v", err)
+	}
+
+	templateNvIpamConfig, err := template.New("nv-ipam-config").Parse(nvIpamConfigTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("template parse error: %v", err)
+	}
+
+	templateData := map[string]string{
+		"KubeConfigFile": o.NvIpamKubeConfigFileHost,
+		"NvIpamDataDir":  o.NvIpamCNIDataDirHost,
+		"NvIpamLogFile":  o.NvIpamLogFile,
+		"NvIpamLogLevel": o.NvIpamLogLevel,
+	}
+
+	// Prepare
+	hash := sha256.New()
+	writer := io.MultiWriter(hash, fp)
+
+	// genearate nv-ipam-config from template
+	if err = templateNvIpamConfig.Execute(writer, templateData); err != nil {
+		return nil, fmt.Errorf("cannot create nv-ipam-config: %v", err)
+	}
+
+	if err := fp.Sync(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot flush nv-ipam-config temp file: %v", err)
+	}
+	if err := fp.Close(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot close nv-ipam-config temp file: %v", err)
+	}
+
+	newFileHash := hash.Sum(nil)
+	if currentFileHash != nil && bytes.Equal(newFileHash, currentFileHash) {
+		fmt.Printf("nv-ipam-config is same, not copy\n")
+		os.Remove(fp.Name())
+		return currentFileHash, nil
+	}
+
+	// replace file with tempfile
+	if err := os.Rename(tempNvIpamConfigFile, nvIpamConfigFile); err != nil {
+		return nil, fmt.Errorf("cannot replace %q with temp file %q: %v", tempNvIpamConfigFile, nvIpamConfigFile, err)
+	}
+
+	fmt.Printf("nv-ipam-config is created in %s\n", nvIpamConfigFile)
```

Comment:
can you add logging for daemon ?
then you can:
1. not use printf
2. emit logs for progress / other useful info

Reply:
technically you can just replace fmt.Printf with log.Printf and remove the ignore in .golangci.yaml

https://pkg.go.dev/log

it will print by default to stderr and will be visible in kubectl logs


later on i might submit PR to switch to a structured logger if there is time.

Reply:
will do